### PR TITLE
OpenAPI 3.0.0 specification support

### PR DESCRIPTION
**Pull request by tgtshanika**
### Proposed changes in this pull request

This PR contains the feature implementation of OpenAPI 3.0.0 support in API Manager 2.2.0. Following are the functionalities enabled with this feature.

- Existing Open API 3.0.0 API definition file/url import in API Manager
- Designing Open API 3.0.0 spec supported APIs.
- Open API 3.0.0 support for Swagger editor
- Open API 3.0.0 support for Swagger console
- REST API support to create/update/retrieve API definition of OpenAPI spec

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-design/js/api-design.js [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
+    "openapi": "3.0.0",
+    "paths": {},
+    "info": {
+        "title": "",
+        "version": ""
+    }
+};
+
+var supportedSwaggerVersion = "2.0";
+var supportedOpenAPI3Version = "3.0.0";
```
Comment (tgtshanika) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-design/js/api-design.js]: Make these constants

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-design/template.jag [hunk 2]
```diff
@@ -874,7 +924,20 @@
...
                 });
-        <% } %>
-    });
-</script>    
+        <% }
+            var apiDefinitionVersion = session.get("apiDefinitionVersion");
+            if(apiDefinitionVersion){
+                %>
+              $.get( jagg.site.context + "/site/blocks/item-design/ajax/add.jag", { "action" : "getAPIDefinitionVersion"} , function( data ) {
+                    if(data != null && data != ""){
```
Comment (ChamNDeSilva) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-design/template.jag]: formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-design/template.jag [hunk 3]
```diff
@@ -874,7 +924,20 @@
 	                    }
 	                }
                 });
-        <% } %>
-    });
-</script>    
+        <% }
+            var apiDefinitionVersion = session.get("apiDefinitionVersion");
+            if(apiDefinitionVersion){
```
Comment (ChamNDeSilva) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/themes/wso2/templates/item-design/template.jag]: formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/wso2/templates/sdk/template.jag [hunk 4]
```diff
@@ -4,6 +4,14 @@
             <% var supportedLangs = jagg.module("sdk").getSupportedLanguages();
                 var langsArray = supportedLangs.langs.split(",");
                 var user = jagg.getUser();
+                if(inputs.api.isOpenAPI3APIDefinition){
```
Comment (ChamNDeSilva) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/store/site/themes/wso2/templates/sdk/template.jag]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 5]
```diff
@@ -1864,6 +1865,18 @@ public static NativeArray jsFunction_getAPI(Context cx, Scriptable thisObj, Obje
                 }
 
                 api = apiConsumer.getAPI(apiIdentifier);
+                String apiDefinition = apiConsumer.getOpenAPIDefinition(apiIdentifier);
+
+                boolean isOpenAPI3APIDefinition = false;
+                if(!apiDefinition.isEmpty()) {
```
Comment (ChamNDeSilva) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIProviderImplTest.java [hunk 6]
```diff
@@ -1733,15 +1733,15 @@ public Void answer(InvocationOnMock invocation) throws Throwable {
         PowerMockito.when(APIUtil.getAPI(newArtifact, apiProvider.registry, api.getId(), "test")).thenReturn(newApi);
 
         //Swagger resource
-        String resourcePath = APIUtil.getSwagger20DefinitionFilePath(api.getId().getApiName(),
+        String resourcePath = APIUtil.getOpenAPIDefinitionFilePath(api.getId().getApiName(),
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIProviderImplTest.java]: formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/APIDefinitionFromOpenAPISpecTest.java [hunk 7]
```diff
@@ -0,0 +1,276 @@
+/*
+ *  Copyright (c) 2005-2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/APIDefinitionFromOpenAPISpecTest.java]: wrong header

**Pull request merged**